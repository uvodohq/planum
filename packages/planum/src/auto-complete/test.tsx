import {
  autoUpdate,
  size,
  useDismiss,
  useFloating,
  useId,
  useInteractions,
  useListNavigation,
  useRole,
} from '@floating-ui/react-dom-interactions'
import React, { forwardRef, useLayoutEffect, useRef, useState } from 'react'

interface ItemProps {
  children: React.ReactNode
  active: boolean
}

const Item = forwardRef<
  HTMLDivElement,
  ItemProps & React.HTMLProps<HTMLDivElement>
>(({ children, active, ...rest }, ref) => {
  const id = useId()
  return (
    <div
      ref={ref}
      role="option"
      id={id}
      aria-selected={active}
      {...rest}
      style={{
        background: active ? 'lightblue' : 'none',
        padding: 4,
        cursor: 'default',
        ...rest.style,
      }}>
      {children}
    </div>
  )
})
Item.displayName = 'Item'

export function AutoCompleteTest({ data }) {
  const [open, setOpen] = useState(false)
  const [inputValue, setInputValue] = useState('')
  const [activeIndex, setActiveIndex] = useState<number | null>(null)

  const listRef = useRef<Array<HTMLElement | null>>([])

  const { x, y, reference, floating, strategy, context, refs } =
    useFloating<HTMLInputElement>({
      whileElementsMounted: autoUpdate,
      open,
      onOpenChange: setOpen,
      middleware: [
        size({
          apply({ rects, elements }) {
            Object.assign(elements.floating.style, {
              width: `${rects.reference.width}px`,
              maxHeight: `${200}px`,
            })
          },
          padding: 10,
        }),
      ],
    })

  useLayoutEffect(() => {
    // IMPORTANT: When the floating element first opens, this runs when the
    // styles have **not yet** been applied to the element. This can cause an
    // infinite loop as `size` has not yet limited the maxHeight, so the whole
    // page tries to scroll. We must wrap it in rAF.
    requestAnimationFrame(() => {
      if (activeIndex != null) {
        listRef.current[activeIndex]?.scrollIntoView({ block: 'nearest' })
      }
    })
  }, [activeIndex])

  const { getReferenceProps, getFloatingProps, getItemProps } = useInteractions(
    [
      useRole(context, { role: 'listbox' }),
      useDismiss(context),
      useListNavigation(context, {
        listRef,
        activeIndex,
        onNavigate: setActiveIndex,
        virtual: true,
        loop: true,
        focusItemOnHover: false,
      }),
    ],
  )

  function onChange(event: React.ChangeEvent<HTMLInputElement>) {
    const value = event.target.value
    setInputValue(value)

    if (value) {
      setOpen(true)
      setActiveIndex(0)
    } else {
      setOpen(false)
    }
  }

  const items = data.filter((item) =>
    item.toLowerCase().startsWith(inputValue.toLowerCase()),
  )

  return (
    <>
      <input
        {...getReferenceProps({
          ref: reference,
          onChange,
          value: inputValue,
          placeholder: 'Enter fruit',
          'aria-autocomplete': 'list',
          onKeyDown(event) {
            if (
              event.key === 'Enter' &&
              activeIndex != null &&
              items[activeIndex]
            ) {
              setInputValue(items[activeIndex])
              setActiveIndex(null)
              setOpen(false)
            }

            if (event.key === 'Tab') {
              setOpen(false)
            }
          },
        })}
      />
      {open && (
        <div
          {...getFloatingProps({
            ref: floating,
            style: {
              position: strategy,
              left: x ?? 0,
              top: y ?? 0,
              background: '#eee',
              color: 'black',
              overflowY: 'auto',
            },
          })}>
          {items.map((item, index) => (
            // eslint-disable-next-line react/jsx-key
            <Item
              {...getItemProps({
                key: item,
                ref(node) {
                  listRef.current[index] = node
                },
                onClick() {
                  setInputValue(item)
                  setOpen(false)
                  refs.reference.current?.focus()
                },
              })}
              active={activeIndex === index}>
              {item}
            </Item>
          ))}
        </div>
      )}
    </>
  )
}
