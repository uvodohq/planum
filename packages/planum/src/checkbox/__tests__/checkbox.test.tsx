import * as React from 'react'
import { vi } from 'vitest'

import { render, screen, userEvent } from '../../test'
import { Checkbox } from '..'

const labelText = 'check me'

describe('Checkbox', async () => {
  it('should render simple checkbox', () => {
    const { container } = render(<Checkbox />)
    const checkbox = screen.getByRole('checkbox')
    expect(checkbox).toBeInTheDocument()
    expect(container).toMatchSnapshot()
  })

  it('with label', () => {
    const { container } = render(<Checkbox label={labelText} />)
    const label = screen.getByText(labelText)
    expect(label).toBeInTheDocument()
    expect(container).toMatchSnapshot()
  })

  it('isDisabled', () => {
    const { container } = render(<Checkbox label={labelText} isDisabled />)
    expect(container).toMatchSnapshot()
  })

  it('isSelected', () => {
    const { container } = render(<Checkbox label={labelText} isSelected />)
    expect(container).toMatchSnapshot()
  })

  it('isIndeterminate', () => {
    const { container } = render(<Checkbox label={labelText} isSelected />)
    expect(container).toMatchSnapshot()
  })

  it('should call onChange', async () => {
    const onChangeMock = vi.fn()
    const { container } = render(<Checkbox onChange={onChangeMock} />)
    const label = container.querySelector('label')!
    await userEvent.click(label)
    expect(onChangeMock).toHaveBeenCalledTimes(1)
    expect(onChangeMock).toHaveBeenCalledWith(true)

    await userEvent.click(label)
    expect(onChangeMock).toHaveBeenCalledWith(false)
  })

  it.skip('should not call onChange when isDisabled', async () => {
    const onChangeMock = vi.fn()
    const { container } = render(
      <Checkbox isDisabled onChange={onChangeMock} />,
    )
    const label = container.querySelector('label')!
    await userEvent.click(label)
    expect(onChangeMock).not.toHaveBeenCalled()
  })

  it('onChange should work with when focused with tab and space key press', async () => {
    const onChangeMock = vi.fn()
    render(<Checkbox onChange={onChangeMock} />)
    const checkbox = screen.getByRole('checkbox')

    await userEvent.tab()
    expect(checkbox).toHaveFocus()
    await userEvent.type(checkbox, '{space}')

    expect(onChangeMock).toHaveBeenCalledWith(true)
  })
})
