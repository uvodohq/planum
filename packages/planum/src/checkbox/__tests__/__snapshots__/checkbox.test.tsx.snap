// Vitest Snapshot v1

exports[`Checkbox > isDisabled 1`] = `
<div>
  <label
    class="c-iPYhUp c-kLRYwF c-edbEbA c-edbEbA-ietuGy-isDisabled-true"
  >
    <div
      style="border: 0px; clip-path: inset(50%); height: 1px; margin: 0px -1px -1px 0px; overflow: hidden; padding: 0px; position: absolute; width: 1px; white-space: nowrap;"
    >
      <input
        aria-checked="false"
        disabled=""
        type="checkbox"
      />
    </div>
    <div
      class="c-hyHTE c-hyHTE-koxIsH-isSelected-false c-hyHTE-hxNqoz-isDisabled-true"
    >
      <span
        class="c-filUNh"
      >
        <svg
          height="14"
          viewBox="0 0 256 256"
          width="14"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h256v256H0z"
            fill="none"
          />
          <path
            d="M216 72 104 184l-56-56"
            fill="none"
            stroke="currentColor"
            stroke-dasharray="400"
            stroke-dashoffset="-400"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="28"
            style="transition: all .2s;"
          />
        </svg>
      </span>
    </div>
    <p
      class="c-iPYhUp c-kLRYwF c-iPYhUp-ikmBbln-css"
    >
      check me
    </p>
  </label>
</div>
`;

exports[`Checkbox > isIndeterminate 1`] = `
<div>
  <label
    class="c-iPYhUp c-kLRYwF c-edbEbA"
  >
    <div
      style="border: 0px; clip-path: inset(50%); height: 1px; margin: 0px -1px -1px 0px; overflow: hidden; padding: 0px; position: absolute; width: 1px; white-space: nowrap;"
    >
      <input
        aria-checked="true"
        checked=""
        type="checkbox"
      />
    </div>
    <div
      class="c-hyHTE c-hyHTE-hkrvYr-isSelected-true"
    >
      <span
        class="c-filUNh"
      >
        <svg
          height="14"
          viewBox="0 0 256 256"
          width="14"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h256v256H0z"
            fill="none"
          />
          <path
            d="M216 72 104 184l-56-56"
            fill="none"
            stroke="currentColor"
            stroke-dasharray="400"
            stroke-dashoffset="0"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="28"
            style="transition: all .2s;"
          />
        </svg>
      </span>
    </div>
    <p
      class="c-iPYhUp c-kLRYwF c-iPYhUp-ikmBbln-css"
    >
      check me
    </p>
  </label>
</div>
`;

exports[`Checkbox > isSelected 1`] = `
<div>
  <label
    class="c-iPYhUp c-kLRYwF c-edbEbA"
  >
    <div
      style="border: 0px; clip-path: inset(50%); height: 1px; margin: 0px -1px -1px 0px; overflow: hidden; padding: 0px; position: absolute; width: 1px; white-space: nowrap;"
    >
      <input
        aria-checked="true"
        checked=""
        type="checkbox"
      />
    </div>
    <div
      class="c-hyHTE c-hyHTE-hkrvYr-isSelected-true"
    >
      <span
        class="c-filUNh"
      >
        <svg
          height="14"
          viewBox="0 0 256 256"
          width="14"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h256v256H0z"
            fill="none"
          />
          <path
            d="M216 72 104 184l-56-56"
            fill="none"
            stroke="currentColor"
            stroke-dasharray="400"
            stroke-dashoffset="0"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="28"
            style="transition: all .2s;"
          />
        </svg>
      </span>
    </div>
    <p
      class="c-iPYhUp c-kLRYwF c-iPYhUp-ikmBbln-css"
    >
      check me
    </p>
  </label>
</div>
`;

exports[`Checkbox > should render simple checkbox 1`] = `
<div>
  <label
    class="c-iPYhUp c-kLRYwF c-edbEbA"
  >
    <div
      style="border: 0px; clip-path: inset(50%); height: 1px; margin: 0px -1px -1px 0px; overflow: hidden; padding: 0px; position: absolute; width: 1px; white-space: nowrap;"
    >
      <input
        aria-checked="false"
        type="checkbox"
      />
    </div>
    <div
      class="c-hyHTE c-hyHTE-koxIsH-isSelected-false"
    >
      <span
        class="c-filUNh"
      >
        <svg
          height="14"
          viewBox="0 0 256 256"
          width="14"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h256v256H0z"
            fill="none"
          />
          <path
            d="M216 72 104 184l-56-56"
            fill="none"
            stroke="currentColor"
            stroke-dasharray="400"
            stroke-dashoffset="-400"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="28"
            style="transition: all .2s;"
          />
        </svg>
      </span>
    </div>
  </label>
</div>
`;

exports[`Checkbox > with label 1`] = `
<div>
  <label
    class="c-iPYhUp c-kLRYwF c-edbEbA"
  >
    <div
      style="border: 0px; clip-path: inset(50%); height: 1px; margin: 0px -1px -1px 0px; overflow: hidden; padding: 0px; position: absolute; width: 1px; white-space: nowrap;"
    >
      <input
        aria-checked="false"
        type="checkbox"
      />
    </div>
    <div
      class="c-hyHTE c-hyHTE-koxIsH-isSelected-false"
    >
      <span
        class="c-filUNh"
      >
        <svg
          height="14"
          viewBox="0 0 256 256"
          width="14"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h256v256H0z"
            fill="none"
          />
          <path
            d="M216 72 104 184l-56-56"
            fill="none"
            stroke="currentColor"
            stroke-dasharray="400"
            stroke-dashoffset="-400"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="28"
            style="transition: all .2s;"
          />
        </svg>
      </span>
    </div>
    <p
      class="c-iPYhUp c-kLRYwF c-iPYhUp-ikmBbln-css"
    >
      check me
    </p>
  </label>
</div>
`;
